docs/proposals/helm-charts.md

```diff
@@ -0,0 +1,206 @@
+a `helm` registry type and have the Helm charts hosted in that Helm chart
+repository converted into objects the broker knows how to manage for use via
+the Service Catalog.
+
+## Problem Description
+
+Currently, our broker only supports docker container image type registries.
+However, with [Helm](https://helm.sh) being the "package manager for Kubernetes",
+there is already a strong collection of applications defined in the form of
+Helm charts. Modifying the broker to consume these Helm charts as Service
+Bundles allows work already done in Helm charts to be made available in
+a cluster.
+
+## Creating a Base Image for Executing Helm Charts
+
+Ansible Playbook Bundles (APBs) have already been described as an instance of
+the more generic Service Bundle. Here, we will be creating a new type of
+Service Bundle, Helm Bundles, that respect the Service Bundle contract but use
+Helm as the runtime. We already have a [Helm Bundle
+Base](https://github.com/ansibleplaybookbundle/helm-bundle-base) and this
+[PR](https://github.com/ansibleplaybookbundle/helm-bundle-base/pull/2)
+shows the work required to implement the Service Bundle contract generically
+using Helm.
+
+## Broker Modifications
+
+### Broker Config Options
+
+It is expected that a broker admin who wants to expose Helm charts would update
+the broker's config to look something like:
+
+```
+registry:
+  - type: helm
+    name: stable
+    url: "https://kubernetes-charts.storage.googleapis.com"
+    base_image: "docker.io/djzager/helm-bundle-base:latest"
+    white_list:
+      - ".*"
+```
+
+1. Type: refers to the registry adapter to handle this registry
+1. Name: gives a name to this registry item
+1. URL: the URL for the Helm chart repository
+1. Base Image: the container image to use when provisioning/deprovisioning a
+   Helm chart
+1. White List/Black List: allows the Admin to filter out Helm charts
+
+### Changes to the Registry Package
+
+- Add the ability to read the `base_image` from the broker config.
+- Pass the URL and BaseImage to the registry adapter.
+- Instantiate a Helm registry adapter to handle Helm Chart registries in the
+  config.
+
+### Creating a Helm Adapter
+
+The `HelmAdapter` struct will look something like the example below:
+
+```
+// import "k8s.io/helm/pkg/repo"
+// HelmAdapter - Helm Registry Adapter
+type HelmAdapter struct {
+    Config Configuration
+    Log    *logging.Logger
+    Charts map[string][]*repo.ChartVersion
+}
+```
+
+The addition of the `Charts` field makes it possible to save our work and only
+read the registries' index file once.
+
+When the broker calls `GetImageNames()` the Helm adapter will read the
+`index.yaml` found at the Helm Chart Repository URL based on the [chart
+repository
+structure](https://github.com/kubernetes/helm/blob/master/docs/chart_repository.md#the-chart-repository-structure).
+The broker will create an entry in the `Charts` field to hold a list of
```

Comment:
Just want to make sure "The broker" here isn't referring to the registry. We definitely don't want generic registry code doing helm specific stuff like writing to the chart cache, that would break the adapter abstraction. I'm thinking it should be the adapter before returning the image name list to the registry.

Also, these adapter methods are going to be called on every bootstrap. What's the condition that we're checking to prevent doing all the "work" involved with each bootstrap to load the cache? i.e.

```go
if cacheNotLoaded() { // What's this?
    adapter.loadCache()
}
```

Reply:
Also, does the value of the chart's entry start out as an empty list prior to the `FetchSpecs` call?

Reply:
Updating to make it clear that the "the broker" is actually the registry. The `Charts` field will be on the `HelmAdapter` struct. Calls to `GetImageNames()` will create the Charts map and `FetchSpecs` will use that map to create the Specs.